### PR TITLE
Add pubsub package

### PR DESCRIPTION
**Pull request by danilotsr**
It adds a simple channel-based publish/subscribe implementation.

### CODE REVIEW

**Comment by hamiltop** [comment]
What's this used for?

**Comment by sumeet** [comment]
nice, this seems pretty useful for in-process signaling. also curious what it will be used for!

File: pubsub/pubsub.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	}
+}
+
+// A single Publisher instance is able to take several subscribers.
+// It returns a receive-only channel.
+func (pub *Publisher) Subscribe() <-chan *Event {
+	ch := make(chan *Event)
+	pub.Lock()
+	pub.subChannels = append(pub.subChannels, ch)
+	pub.Unlock()
```
Comment (sumeet) [inline pubsub/pubsub.go]: shouldn't this also be a `defer pub.Unlock()`?

File: pubsub/pubsub.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+func NewPublisher() *Publisher {
+	return &Publisher{
+		subChannels: []chan *Event{},
+	}
+}
+
+// A single Publisher instance is able to take several subscribers.
+// It returns a receive-only channel.
+func (pub *Publisher) Subscribe() <-chan *Event {
+	ch := make(chan *Event)
```
Comment (bmarini) [inline pubsub/pubsub.go]: This is an unbuffered channel, might want to allow the subscriber to choose the amount of buffer via an arg to `Subscribe`

**Comment by bmarini** [comment]
@hamiltop @sumeet this package is being used in hermes right now, @danilotsr is extracting it out here.

**Comment by danilotsr** [comment]
This package is being used in r101-hermes, see https://github.com/remind101/r101-hermes/pull/432.

**Comment by Kevin-Jonaitis** [comment]
As per r2d2, closing this as this person no longer works here.

**Pull request closed without merging**